app/services/work_packages/set_attributes_service/derive_progress_values_base.rb

```diff
@@ -0,0 +1,153 @@
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+class WorkPackages::SetAttributesService
+  class DeriveProgressValuesBase
+    attr_reader :work_package
+
+    PROGRESS_ATTRIBUTES = %i[work remaining_work percent_complete].freeze
+
+    def initialize(work_package)
+      @work_package = work_package
+    end
+
+    def call
+      derive_progress_attributes
+      round_progress_values
+    end
+
+    def work
+      work_package.estimated_hours
+    end
+
+    def work=(value)
+      work_package.estimated_hours = value
+    end
+
+    def work_was
+      work_package.estimated_hours_was
+    end
+
+    def work_set?
+      work.present?
+    end
+
+    def work_unset?
+      work.nil?
+    end
+
+    def work_was_unset?
+      work_package.estimated_hours_was.nil?
+    end
+
+    def work_changed?
+      work_package.estimated_hours_changed?
+    end
+
+    def work_came_from_user?
+      work_package.estimated_hours_came_from_user?
+    end
+
+    def remaining_work
+      work_package.remaining_hours
+    end
+
+    def remaining_work=(value)
+      work_package.remaining_hours = value
+    end
+
+    def remaining_work_set?
+      remaining_work.present?
+    end
+
+    def remaining_work_unset?
+      remaining_work.nil?
+    end
+
+    def remaining_work_was_unset?
+      work_package.remaining_hours_was.nil?
+    end
+
+    def remaining_work_changed?
+      work_package.remaining_hours_changed?
+    end
+
+    def remaining_work_came_from_user?
+      work_package.remaining_hours_came_from_user?
+    end
+
+    def percent_complete
+      work_package.done_ratio
+    end
+
+    def percent_complete=(value)
+      work_package.done_ratio = value
+    end
+
+    def percent_complete_set?
+      percent_complete.present?
+    end
+
+    def percent_complete_unset?
+      percent_complete.nil?
+    end
+
+    def percent_complete_was_unset?
+      work_package.done_ratio_was.nil?
+    end
+
+    def percent_complete_changed?
+      work_package.done_ratio_changed?
+    end
+
+    def percent_complete_came_from_user?
+      work_package.done_ratio_came_from_user?
+    end
+
+    private
+
+    def round_progress_values
+      rounded = work&.round(2)
+      if rounded != work
```

Comment:
Does this prevent some change detection bug?

Reply:
Yes! Thankfully there is a test failing when I remove the condition.

If `estimated_hours` is set to a string like `"Hello world!"`, thenactiverecord will coerce and the value to `0.0`. When doing the rounding, it will set it to `0.0` again. So what is the difference?

The difference is the value returned by `work_package.estimated_hours_before_type_cast`: it will be `"Hello world!"` first, but as soon as the value is set to `0.0` by the rounding, `work_package.estimated_hours_before_type_cast` will return `0.0` too, and the numericality validator (`validates :estimated_hours, numericality: true`) on the model will fail to detect that the value was incorrectly set to a string.

Reply:
I added a comment about it.